src/frontend/src/optimizer/rule/distinct_agg_rule.rs

```diff
@@ -29,13 +29,23 @@ use crate::optimizer::PlanRef;
 use crate::utils::{ColIndexMapping, Condition};
 
 /// Transform distinct aggregates to `LogicalAgg` -> `LogicalAgg` -> `Expand` -> `Input`.
-pub struct DistinctAggRule {}
+pub struct DistinctAggRule {
+    for_stream: bool,
+}
+
 impl Rule for DistinctAggRule {
     fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
         let agg: &LogicalAgg = plan.as_logical_agg()?;
         let (mut agg_calls, mut agg_group_keys, input) = agg.clone().decompose();
         let original_group_keys_len = agg_group_keys.len();
 
+        if self.for_stream {
+            // Due to performance issue, we don't do 2-phase agg for stream distinct agg with group
+            // by. See https://github.com/risingwavelabs/risingwave/issues/7271 for more.
+            // TODO(rc): may be we can still apply the rule if there is only one distinct column
+            return None;
+        }
```

Comment:
I think we should still apply this rule if the agg matched here without group by, right?

Reply:
should we also apply the distinct aggregator for group Aggregation :thinking: ?

Reply:
3 cases here:

- group by + multiple distinct columns => disable this rule will make 50% boost
- no group by + multiple distinct columns => disable this rule will make 22% boost
- single distinct column (can be optimized to group agg) => still need evaluation

I'll do some test for the 3rd case today.

Reply:
Simple evaluation on my local machine shows that it has a 30+% boost in the third case. 

But since using backend distinct agg impl introduced in this PR may harm scalability, I've modified the code to maintain the old behavior for the last 2 cases.